### PR TITLE
Power analysis for t test, ANOVA, Chi-Square

### PR DESCRIPTION
**Pull request by hideaki**
# Description
Power analysis for t test, ANOVA, Chi-Square.

# Checklist
Make sure you have performed following items before submitting this pull request.
If not, please describe the reason.  

- [x] Add test cases for this fix/enhancement
- [ ] Pass devtools::check()
- [ ] Pass devtools::test()
- [ ] Test installing from github
- [x] Tested with Exploratory

### CODE REVIEW

File: R/test_wrapper.R [hunk 1]
```diff
@@ -348,13 +367,53 @@ tidy.chisq_exploratory <- function(x, type = "observed") {
 glance.chisq_exploratory <- function(x) {
   # ret <- x %>% broom:::glance.htest() # for some reason this does not work. just do it like following.
   ret <- data.frame(statistic=x$statistic, parameter=x$parameter, p.value=x$p.value)
-  ret <- ret %>% rename(`Chi-Square`=statistic, `Degree of Freedom`=parameter, `P Value`=p.value)
+  N <- sum(x$observed) # Total number of observations (rows).
+  if (is.null(x$power)) {
+    # If power is not specified in the arguments, estimate current power.
+    # x$parameter is degree of freedom.
+    if (!is.na(x$cohens_w_to_detect)) { # It is possible that x$cohens_w_to_detect is NA. Avoid calling pwr.chisq.test not to hit an error.
+      power_res <- pwr::pwr.chisq.test(df = x$parameter, N = N, w = x$cohens_w_to_detect, sig.level = x$sig.level)
```
Comment (hideaki) [inline R/test_wrapper.R]: enclose with tryCatch() to handle other errors too.

File: R/test_wrapper.R [hunk 2]
```diff
@@ -348,13 +367,53 @@ tidy.chisq_exploratory <- function(x, type = "observed") {
...
+    ret <- ret %>% rename(`Chi-Square`=statistic,
+                          `Degree of Freedom`=parameter,
+                          `P Value`=p.value,
+                          `Effect Size (Cohen's w)`=w,
+                          `Power`=power,
+                          `Probability of Type 2 Error`=beta)
+  }
+  else {
+    # If required power is specified in the arguments, estimate required sample size. 
+    power_res <- pwr::pwr.chisq.test(df = x$parameter, w = x$cohens_w_to_detect, sig.level = x$sig.level, power = x$power)
```
Comment (hideaki) [inline R/test_wrapper.R]: Add comment on how this function works (e.g. calculates missing one)

Comment (hideaki) [reply R/test_wrapper.R]: Enclose with tryCatch().

File: R/test_wrapper.R [hunk 3]
```diff
@@ -430,13 +515,53 @@ tidy.ttest_exploratory <- function(x, type="model", conf_level=0.95) {
...
-                    `P Value`=p.value,
-                    `Degree of Freedom`=parameter,
-                    Difference=estimate,
-                    `Conf High`=conf.high,
-                    `Conf Low`=conf.low)
+
+    data_summary <- x$data %>% dplyr::group_by(!!rlang::sym(x$var2)) %>%
+      dplyr::summarize(n_rows=length(!!rlang::sym(x$var1)))
+    n1 <- data_summary$n_rows[[1]]
+    n2 <- data_summary$n_rows[[2]]
```
Comment (hideaki) [inline R/test_wrapper.R]: Add comment on what they are.

File: R/test_wrapper.R [hunk 4]
```diff
@@ -535,13 +676,48 @@ glance.anova_exploratory <- function(x) {
...
     ret <- broom:::tidy.aov(x)
-    ret <- ret %>% dplyr::select(term, statistic, p.value, df, sumsq, meansq) %>%
-      dplyr::rename(Term=term,
-                    `F Ratio`=statistic,
-                    `P Value`=p.value,
-                    `Degree of Freedom`=df,
-                    `Sum of Squares`=sumsq,
-                    `Mean Square`=meansq)
+
+    data_summary <- x$data %>% dplyr::group_by(!!rlang::sym(x$var2)) %>%
```
Comment (hideaki) [inline R/test_wrapper.R]: Add comment.

File: R/test_wrapper.R [hunk 5]
```diff
@@ -535,13 +676,48 @@ glance.anova_exploratory <- function(x) {
...
+      dplyr::summarize(n_rows=length(!!rlang::sym(x$var1))) %>%
+      dplyr::summarize(min_n_rows=min(n_rows), k=n())
+    k <- data_summary$k
```
Comment (hideaki) [inline R/test_wrapper.R]: Add comment on what k is.

File: R/util.R [hunk 6]
```diff
@@ -1635,4 +1635,69 @@ setdiff <- function(x, y, force_data_type = FALSE, ...){
   }
 }
 
+# This is written by removing unnecessary part from calculate_cohens_d.
+#'Calculate common standard deviation.
+#'@export
+calculate_common_sd <- function(var1, var2) {
+  df <- data.frame(var1=var1, var2=var2)
+  summarized <- df %>% dplyr::group_by(var2) %>%
+    dplyr::summarize(n=n(), v=var(var1))
```
Comment (hideaki) [inline R/util.R]: Add na.rm = TRUE

File: R/util.R [hunk 7]
```diff
@@ -1635,4 +1635,69 @@ setdiff <- function(x, y, force_data_type = FALSE, ...){
...
+  csd <- sqrt(csd) # common sd computation
+}
+
+# Reference: https://stackoverflow.com/questions/15436702/estimate-cohens-d-for-effect-size
+#'Calculate Cohen's d
+#'@export
+calculate_cohens_d <- function(var1, var2) {
+  df <- data.frame(var1=var1, var2=var2)
+  summarized <- df %>% dplyr::group_by(var2) %>%
+    dplyr::summarize(n=n(), m=mean(var1), v=var(var1))
```
Comment (hideaki) [inline R/util.R]: Add na.rm = TRUE

Comment (hideaki) [reply R/util.R]: done.

File: R/util.R (hunk 6 above)
Comment (hideaki) [reply R/util.R]: done.

File: R/test_wrapper.R (hunk 5 above)
Comment (hideaki) [reply R/test_wrapper.R]: done.

File: R/test_wrapper.R (hunk 3 above)
Comment (hideaki) [reply R/test_wrapper.R]: done.

File: R/test_wrapper.R (hunk 4 above)
Comment (hideaki) [reply R/test_wrapper.R]: done.

File: R/util.R [hunk 8]
```diff
@@ -1635,6 +1635,72 @@ setdiff <- function(x, y, force_data_type = FALSE, ...){
...
+# Did not use sjstats::cohens_f() to avoid requiring entire sjstats and its dependencies.
+# Did not use powerAnalysis::ES.anova.oneway() because it only works for the case all categories
+# have same number of observations.
+#'Calculate Cohen's f
+#'@export
+calculate_cohens_f <- function(var1, var2) {
+  m <- mean(var1, na.rm = TRUE)
+  df <- data.frame(var1=var1, var2=var2)
+  summarized <- df %>% dplyr::group_by(var2) %>%
+    dplyr::mutate(diff_between = mean(var1) - m, diff_total = var1 - m) %>% dplyr::ungroup() %>%
```
Comment (kei51e) [inline R/util.R]: Let's add na.rm=TRUE for mean and sum too.

Comment (hideaki) [reply R/util.R]: Done.

**Pull request merged**